flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,217 @@
+        DataStream<Row> input = tEnv.toDataStream(inputs[0]);
+        String labelCol = getLabelCol();
+
+        DataStream<Tuple2<DenseVector, Integer>> trainData =
+                input.map(
+                        new MapFunction<Row, Tuple2<DenseVector, Integer>>() {
+                            @Override
+                            public Tuple2<DenseVector, Integer> map(Row value) {
+                                Integer label = (Integer) value.getField(labelCol);
+                                DenseVector vec = (DenseVector) value.getField(getFeaturesCol());
+                                return Tuple2.of(vec, label);
+                            }
+                        });
+
+        DataStream<Row> model = buildModel(trainData);
+        KnnModel knnModel =
+                new KnnModel()
+                        .setFeaturesCol(getFeaturesCol())
+                        .setK(getK())
+                        .setPredictionCol(getPredictionCol());
+        knnModel.setModelData(tEnv.fromDataStream(model, KnnModelData.getModelSchema()));
+        return knnModel;
+    }
+
+    /**
+     * Builds knn model.
+     *
+     * @param dataStream Input data.
+     * @return Knn model.
+     */
+    private static DataStream<Row> buildModel(DataStream<Tuple2<DenseVector, Integer>> dataStream) {
+        Schema schema = KnnModelData.getModelSchema();
+        return dataStream.transform(
+                "build knn model",
+                TableUtils.getRowTypeInfo(schema),
+                new MapPartitionFunctionWrapper<>(
+                        new RichMapPartitionFunction<Tuple2<DenseVector, Integer>, Row>() {
+                            @Override
+                            public void mapPartition(
+                                    Iterable<Tuple2<DenseVector, Integer>> values,
+                                    Collector<Row> out) {
+                                List<Tuple3<DenseMatrix, DenseVector, Integer[]>> list =
+                                        prepareMatrixData(values);
+                                for (Tuple3<DenseMatrix, DenseVector, Integer[]> t3 : list) {
+                                    Row ret = new Row(3);
+                                    ret.setField(0, t3.f0);
+                                    ret.setField(1, t3.f1);
+                                    ret.setField(2, t3.f2);
+                                    out.collect(ret);
+                                }
+                            }
+                        }));
+    }
+
+    /**
+     * Prepares matrix data, the output is a list of Tuple3, which includes vectors, vecNorms and
+     * label.
+     *
+     * @param trainData Input train data.
+     * @return Model data in format of list tuple3.
+     */
+    private static List<Tuple3<DenseMatrix, DenseVector, Integer[]>> prepareMatrixData(
+            Iterable<Tuple2<DenseVector, Integer>> trainData) {
+        final int size = 5 * 1024 * 1024;
```

Comment:
what is the intuition for choose `size=5*1024*1024`?

Reply:
L2 cache in computer architecture is considered when we choose the size. Then we can decrease cache miss rate.